include/exec/system_context.hpp

```diff
@@ -0,0 +1,433 @@
+      stdexec::set_error(std::move(__self.__state_.__recv_), std::move(__ptr));
+    }
+
+    /// Gets the environment of this receiver; returns the environment of the connected receiver.
+    friend decltype(auto)
+      tag_invoke(stdexec::get_env_t, const __bulk_intermediate_receiver& __self) noexcept {
+      return stdexec::get_env(__self.__state_.__recv_);
+    }
+
+    /// Called by the system scheduler when the bulk operation completes.
+    static void __cb(void* __data, int __completion_type, void* __exception) {
+      auto __state = static_cast<__bulk_state_t*>(__data);
+      delete[] __state->__arguments_data_;
+      __detail::__pass_to_receiver(__completion_type, __exception, std::move(__state->__recv_));
+    }
+  };
+
+  /// The operation state object for the system bulk sender.
+  template <stdexec::sender __Previous, std::integral __Size, class __Fn, class __R>
+  struct __bulk_op {
+    /// The inner operation state, which is the result of connecting the previous sender to the bulk intermediate receiver.
+    using __inner_op_state = stdexec::
+      connect_result_t<__Previous, __bulk_intermediate_receiver<__Previous, __Size, __Fn, __R>>;
+
+    /// Constructs `this` from `__snd` and `__recv`, using the object returned by `__initFunc` to start the operation.
+    ///
+    /// Using a functor to initialize the operation state allows the use of `this` to get the
+    /// underlying implementation object.
+    template <class __InitF>
+    __bulk_op(
+      system_bulk_sender<__Previous, __Size, __Fn>&& __snd,
+      __R&& __recv,
+      __InitF&& __initFunc)
+      : __state_{std::move(__snd), std::move(__recv)}
+      , __previous_operation_state_{__initFunc(*this)} {
+    }
+
+    __bulk_op(const __bulk_op&) = delete;
+    __bulk_op(__bulk_op&&) = delete;
+    __bulk_op& operator=(const __bulk_op&) = delete;
+    __bulk_op& operator=(__bulk_op&&) = delete;
+
+    /// Starts the work stored in `__self`.
+    friend void tag_invoke(stdexec::start_t, __bulk_op& __self) noexcept {
+      // Start previous operation state.
+      // Bulk operation will be started when the previous sender completes.
+      stdexec::start(__self.__previous_operation_state_);
+    }
+
+    /// The state of this bulk operation.
+    __bulk_state<__Previous, __Size, __Fn, __R> __state_;
+    /// The operation state object of the previous computation.
+    __inner_op_state __previous_operation_state_;
+  };
+
+  /// The sender used to schedule bulk work in the system context.
+  template <stdexec::sender __Previous, std::integral __Size, class __Fn>
+  struct system_bulk_sender {
+    /// Marks this type as being a sender; not to spec.
+    using sender_concept = stdexec::sender_t;
+    /// Declares the completion signals sent by `this`.
+    using completion_signatures = stdexec::completion_signatures<
+      stdexec::set_value_t(),
+      stdexec::set_stopped_t(),
+      stdexec::set_error_t(std::exception_ptr) >;
```

Comment:
This isn't correct. The `bulk` sender should send the values and errors of its predecessor.

Reply:
Indeed, see Lee's TODO 2 lines below.

However, my understanding that this sender will always complete with `set_value()` with no arguments, regardless of the values sent by the previous senders; these values are just used to pass in to the given user function. Is my understanding wrong?

On the error side, my plan is to forward all errors given by the previous sender, plus `std::exception_ptr`, so that failure to schedule on the external scheduler can be reported as an error.

Reply:
> However, my understanding that this sender will always complete with `set_value()` with no arguments, regardless of the values sent by the previous senders

Not so. `bulk` is specified to move the stored values into the value channel after the bulk work is done.